### PR TITLE
Display Delegation information on DelegationNetwork page.

### PR DESCRIPTION
**Pull request by andorp**

### CODE REVIEW

File: src/Frontend/Page/User.hs [hunk 1]
```diff
@@ -60,12 +60,13 @@ data PageUserProfileCreatedIdeas = PageUserProfileCreatedIdeas CapCtx UserView L
 instance Page PageUserProfileCreatedIdeas where
     isAuthorized = userPage -- Are profiles public?
 
-newtype DelegationInfo = DelegationInfo [(User, [User])]
+-- Represents two step long delegation paths from a user.
+newtype DelegationPath = DelegationPath [(User, [User])]
```
Comment (fisx) [inline src/Frontend/Page/User.hs]: `DelegationTree`

File: src/Frontend/Page/Delegation.hs [hunk 2]
```diff
@@ -80,21 +83,19 @@ topicDelegation tid = formPageHandlerWithMsg
 
 -- | 13. Delegation network
 -- FIXME: Render all the delegations not just the current user related ones.
-data PageDelegationNetwork = PageDelegationNetwork [Delegation]
+data PageDelegationNetwork = PageDelegationNetwork DScope [DScope] [DelegationInfo]
```
Comment (fisx) [inline src/Frontend/Page/Delegation.hs]: also interesting here: which DScopes are parents of whom?

Comment (fisx) [inline src/Frontend/Page/Delegation.hs]: (not urgent, could be done in this PR or in a separate one.)

File: src/Frontend/Page/Delegation.hs [hunk 3]
```diff
@@ -152,7 +153,43 @@ instance ToHtml PageDelegationNetwork where
...
+    type FormPageResult  PageDelegationNetwork = PageDelegationNetworkPayload
+
+    formAction (PageDelegationNetwork scope _ _)      = U.delegationViewScope scope
+    redirectOf _ (PageDelegationNetworkPayload scope) = U.delegationViewScope scope
+
+    makeForm (PageDelegationNetwork actualDScope dscopes _delegations) =
+        PageDelegationNetworkPayload
+        <$> ("scope" .: DF.choice delegationScopeList (Just actualDScope))
+      where
+        -- TODO: Better visual instead of show
```
Comment (fisx) [inline src/Frontend/Page/Delegation.hs]: yes, good idea.  if there is no HasUILabel instance, just write one.

File: src/Types.hs [hunk 4]
```diff
@@ -804,6 +812,24 @@ data DScope =
 
 instance SOP.Generic DScope
 
+readEitherCS :: (ConvertibleStrings String c, Read a) => String -> Either c a
+readEitherCS = either (Left . cs) Right . readEither
```
Comment (fisx) [inline src/Types.hs]: move to mini-prelude in Types.hs?

File: src/Frontend/Page/Delegation.hs [hunk 5]
```diff
@@ -152,7 +153,43 @@ instance ToHtml PageDelegationNetwork where
...
+        form $ do
+            inputSelect_ [] "scope" v
+            DF.inputSubmit "Show delegations!"
+            p_ . toHtml $ show delegations
+            img_ [src_ . U.TopStatic $ "images" </> "delegation_network_dummy.jpg"]
+
+viewDelegationNetwork :: ActionM m => Maybe DScope -> FormPageHandler m PageDelegationNetwork
+viewDelegationNetwork mscope = formPageHandler
+    (do user <- currentUser
+        let scope = fromMaybe DScopeGlobal mscope
```
Comment (fisx) [inline src/Frontend/Page/Delegation.hs]: this is a perfect use case for a view pattern, as `mscope` is not needed.

Comment (fisx) [inline src/Frontend/Page/Delegation.hs]: (just saying :-)

**Comment by fisx** [comment]
LGTM!  for details see comments.

File: exec/RenderHtml.hs [hunk 6]
```diff
@@ -91,6 +90,7 @@ pagesForm f =
     , f (Proxy :: Proxy PageUserSettings)
     , f (Proxy :: Proxy PageAdminSettingsEventsProtocol)
     , f (Proxy :: Proxy PageDelegateVote)
+--    , f (Proxy :: Proxy PageDelegationNetwork) -- FIXME
```
Comment (fisx) [inline exec/RenderHtml.hs]: i commented that back in and it works.  why did you take it out?

File: src/Frontend.hs [hunk 7]
```diff
@@ -159,7 +159,7 @@ type AulaMain =
 
        -- delegation network
   :<|> "delegation" :> "edit" :> FormHandler PageDelegateVote
-  :<|> "delegation" :> "view" :> GetH (Frame PageDelegationNetwork)
+  :<|> "delegation" :> "view" :> QueryParam "score" DScope :> FormHandler PageDelegationNetwork
```
Comment (fisx) [inline src/Frontend.hs]: you still need to fix this typo (`score` => `scope`).

File: exec/RenderHtml.hs (hunk 6 above)
Comment (fisx) [inline exec/RenderHtml.hs]: no need to take it back in, btw.  i've already done this in my PR.  just curious.

Comment (andorp) [inline exec/RenderHtml.hs]: I just inserted there for future work.

**Comment by fisx** [comment]
one more thing: we already have `DelegationNetwork`, which does the same as your new `DelegationInfo`.  either one of them should go, but it's ok if you only write a FIXME about that.

File: exec/RenderHtml.hs (hunk 6 above)
Comment (andorp) [inline exec/RenderHtml.hs]: I am happy if it works without any additional work. Previously it was a Page instance, not a FormPage.

File: src/Frontend/Page/Delegation.hs (hunk 5 above)
Comment (andorp) [inline src/Frontend/Page/Delegation.hs]: Fixed in: 53e9bfd

**Comment by fisx** [comment]
entering the page at dspace idea-1, but no such idea exists, and active menu entry is global.

File: src/Frontend/Page/Delegation.hs [hunk 8]
```diff
@@ -166,14 +166,13 @@ instance FormPage PageDelegationNetwork where
         <$> ("scope" .: DF.choice delegationScopeList (Just actualDScope))
       where
         -- TODO: Better visual instead of show
```
Comment (fisx) [inline src/Frontend/Page/Delegation.hs]: dangling TODO.  i'll remove that.

**Comment by fisx** [comment]
i see only "schule" and "idee" dscopes in the menu.  is that intended?

File: src/Persistent/Pure.hs [hunk 9]
```diff
@@ -606,6 +608,31 @@ addDelegation = addDb dbDelegationMap
 deleteDelegation :: AUID Delegation -> AUpdate ()
 deleteDelegation did = dbDelegationMap . at did .= Nothing
 
+delegationScopeTree :: User -> Query (Tree DScopeFull)
+delegationScopeTree user = unfoldTreeM discover DScopeGlobalFull
+  where
+    discover :: DScopeFull -> Query (DScopeFull, [DScopeFull])
+    discover DScopeGlobalFull = do
+        schoolIdeas <- DScopeIdeaFull <$$> findWildIdeasBySpace SchoolSpace
+        let ideaSpaces = DScopeIdeaSpaceFull . ClassSpace <$> (user ^.. userSchoolClasses)
```
Comment (fisx) [inline src/Persistent/Pure.hs]: moderator should be able to see all classes.  we have that functionality somewhere, let me find it...

Comment (fisx) [inline src/Persistent/Pure.hs]: no, what i meant was `RoleScope` but that's different in that it drops the difference between "SchoolScope" and "all classes".  we need that distinction here.

File: src/Persistent/Pure.hs [hunk 10]
```diff
@@ -606,6 +608,31 @@ addDelegation = addDb dbDelegationMap
...
+        pure (DScopeGlobalFull, schoolIdeas <> ideaSpaces)
+
+    -- FIXME: Impossible: As the Global is the School space
+    discover s@(DScopeIdeaSpaceFull SchoolSpace) = do
```
Comment (fisx) [inline src/Persistent/Pure.hs]: this should be removed, and the next case should cover this (just relax the pattern).

**Comment by fisx** [comment]
LGTM.  will merge when green.

thanks!  (-:

**Comment by fisx** [comment]
Ok, travis now runs 17 mins.  Thanks for this tweak!

**Pull request merged**